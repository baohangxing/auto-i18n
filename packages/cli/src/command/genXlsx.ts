import fsExtra from 'fs-extra'
import { getKeys, getValueByKey } from '@h1mple/auto-i18n-core'
import { getAutoConfig, getJsonPath, getOutputFileDir, isUnTransed } from '../config/config'
import { KEY_SYMBOL_IN_XLSX } from '../config/constants'
import { writeXlsxFile } from '../utils/excel'
import { createFileName } from '../utils/help'

const { readJsonSync } = fsExtra

const generateXlsx = async () => {
  const autoConfig = getAutoConfig()

  const { baseLangJson, otherLangJsons } = getJsonPath()

  const baseLangJsonObj = readJsonSync(baseLangJson.path)

  const baseJsonKeySet = new Set(getKeys(baseLangJsonObj))

  const firstRows: string[] = [KEY_SYMBOL_IN_XLSX]
  const sheetsDatas: string[][] = []

  firstRows.push(baseLangJson.name)

  const langJsonObjMap = {} as any

  for (const lang of otherLangJsons) {
    firstRows.push(lang.name)
    langJsonObjMap[lang.name] = readJsonSync(lang.path)
  }

  for (const key of baseJsonKeySet) {
    const newRow: string[] = []
    newRow.push(key)
    newRow.push(getValueByKey(baseLangJsonObj, key))

    for (const lang of otherLangJsons) {
      const langVal = getValueByKey(langJsonObjMap[lang.name], key)
      newRow.push(isUnTransed(langVal, lang.name) ? '' : langVal)
    }

    sheetsDatas.push(newRow)
  }

  writeXlsxFile([firstRows], ['all'], [sheetsDatas],
    getOutputFileDir(`${createFileName(autoConfig.outputXlsxNameBy.genXlsx)}.xlsx`),
  )
}

export { generateXlsx }
