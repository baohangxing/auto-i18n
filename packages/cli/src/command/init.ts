import path from 'path'
import { cosmiconfigSync } from 'cosmiconfig'
import fs from 'fs-extra'
import serialize from 'serialize-javascript'
import { checkEslintConfigExist, fgSync, lintFiles } from '@yostar/auto-i18n-core'
import { CLI_CONFIG_NAME } from '../config/constants'
import log from '../utils/log'
import defaultAutoConfig from '../config/defaultAutoConfig'

const getBaseLocales = (locales: string[]): string => {
  const mayZhNames = ['zh', 'zh-cn', 'cn', 'zg', 'china']
  for (const locale of locales) {
    for (const name of mayZhNames) {
      if (locale.toLowerCase().includes(name.toLowerCase()))
        return locale
    }
  }
  return ''
}

const init = async () => {
  const explorer = cosmiconfigSync(CLI_CONFIG_NAME)
  const result = explorer.search()

  if (result?.config) {
    log.info('The project had inited, so skipped')
  }
  else {
    const configPath = path.join(process.cwd(), `${CLI_CONFIG_NAME}.config.cjs`)

    const config = { ...defaultAutoConfig }

    const paths = fgSync(config.localesJsonDirs)

    config.locales = [...new Set(paths.map(x => path.parse(x).name))].sort()
    config.baseLocale = getBaseLocales(config.locales)

    if (checkEslintConfigExist())
      config.autoFormat = true

    const code
    = `// @ts-check
/** @typedef {import('@yostar/auto-i18n-cli').AutoConfig} AutoConfig **/

/** @type AutoConfig */
module.exports = ${serialize(config, {
      unsafe: true,
    })}`

    fs.outputFileSync(configPath, code)

    if (config.autoFormat)
      await lintFiles(configPath)
    log.success('Inited')
  }
}

export { init }
