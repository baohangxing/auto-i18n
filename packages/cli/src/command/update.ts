import path from 'path'
import fsExtra from 'fs-extra'
import {
  getKeys, getValueByKey, setValueByKey,
  sortObjectKey,
} from '@yostar/auto-i18n-core'
import {
  writeToJsonFile,
} from '../utils/help'
import { getAutoConfig, getJsonPath, isUnTransed, unTransLocale } from '../config/config'
import log from '../utils/log'

import { readXlsxFile } from '../utils/excel'
import { KEY_SYMBOL_IN_XLSX } from '../config/constants'

const { readJsonSync } = fsExtra

const updateBaseLocale = async (replaceObj?: any) => {
  const { baseLangJson } = getJsonPath()
  const autoConfig = getAutoConfig()

  if (!baseLangJson) {
    log.error(`No ${autoConfig.baseLocale} JSON file in ${autoConfig.localesJsonDirs}`)
    return
  }

  let baseLangJsonObj

  if (replaceObj) {
    baseLangJsonObj = {}
    const addKeySet = new Set(getKeys(replaceObj))

    for (const k of addKeySet) {
      const v = getValueByKey(replaceObj, k)
      setValueByKey(baseLangJsonObj, k, v)
    }
  }
  else {
    baseLangJsonObj = readJsonSync(baseLangJson.path)
  }
  await writeToJsonFile(path.parse(baseLangJson.path).dir, baseLangJson.name, baseLangJsonObj)
}

const updateLocales = async () => {
  const autoConfig = getAutoConfig()
  const { baseLangJson, otherLangJsons } = getJsonPath()

  if (!baseLangJson) {
    log.error(`No ${autoConfig.baseLocale} JSON file in ${autoConfig.localesJsonDirs}`)
    return
  }

  const baseLangJsonObj = readJsonSync(baseLangJson.path)

  await updateBaseLocale()

  for (const langJson of otherLangJsons) {
    const langJsonObj = readJsonSync(langJson.path)
    const baseJsonKeySet = new Set(getKeys(baseLangJsonObj))
    const langJsonKeySet = new Set(getKeys(langJsonObj))
    const newObj = {}
    for (const x of baseJsonKeySet) {
      const baseValue = getValueByKey(baseLangJsonObj, x)
      const langValue = getValueByKey(langJsonObj, x)

      if (langValue === undefined
        || langValue === null
        || langValue === ''
        || isUnTransed(langValue, langJson.name)) {
        let transedWord = ''
        if (autoConfig.transLacaleWord) {
          transedWord = await autoConfig.transLacaleWord(
            baseValue,
            baseLangJson.name,
            langJson.name,
          )
        }
        else {
          transedWord = unTransLocale(baseValue, langJson.name)
        }
        setValueByKey(newObj, x, transedWord)
      }
      else if (typeof langValue === 'string') {
        setValueByKey(newObj, x, langValue)
      }
      langJsonKeySet.delete(x)
    }
    if (langJsonKeySet.size > 0) {
      for (const x of langJsonKeySet) {
        log.info(
          `${x}: ${getValueByKey(langJsonObj, x)} in ${langJson.name
          }.json is deleted, pleace check it is used in the project!`,
        )
      }
    }

    await writeToJsonFile(path.parse(langJson.path).dir, langJson.name, newObj)
  }

  log.success('Update JSON files success')
}

const updateLocalesFromXlsx = async (filePath: string) => {
  const res = readXlsxFile(path.resolve(filePath))

  const { baseLangJson, otherLangJsons } = getJsonPath()
  if (res) {
    if (!baseLangJson) {
      const autoConfig = getAutoConfig()
      log.error(`No ${autoConfig.baseLocale} JSON file in ${autoConfig.localesJsonDirs}`)
      return
    }
    const baseLangJsonObj = readJsonSync(baseLangJson.path)
    for (const langJson of otherLangJsons) {
      const langJsonObj = readJsonSync(langJson.path)
      const typeJsonKeySet = new Set(getKeys(baseLangJsonObj))
      const newJsonObj = sortObjectKey(langJsonObj)

      for (const { data } of res) {
        const keyIndex = Array.isArray(data[0]) ? data[0].indexOf(KEY_SYMBOL_IN_XLSX) : -1
        const langIndex = Array.isArray(data[0]) ? data[0].indexOf(langJson.name) : -1

        if (keyIndex !== -1 && langIndex !== -1) {
          for (const aLine of data.slice(1)) {
            if (Array.isArray(aLine) && typeJsonKeySet.has(aLine[keyIndex])) {
              if (aLine[langIndex]) {
                setValueByKey(
                  newJsonObj,
                  aLine[keyIndex],
                  aLine[langIndex].toString(),
                )
              }
            }
          }
        }
      }
      await writeToJsonFile(path.parse(langJson.path).dir, langJson.name, newJsonObj)
      log.success(`Write to JSON file success in ${langJson.path}`)
    }
  }
}

export { updateLocalesFromXlsx, updateLocales, updateBaseLocale }
