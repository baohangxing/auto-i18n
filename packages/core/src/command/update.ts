import path from 'path'
import fsExtra from 'fs-extra'
import {
  getKeys, getValueByKey, setValueByKey,
  sortObjectKey,
  writeToJsonFile,
} from '../utils/help'
import { getAutoConfig, getJsonPath, isUnTransed, unTransLocale } from '../config/config'
import log from '../utils/log'

import { readXlsxFile } from '../utils/excel'
import { KEY_SYMBOL_IN_XLSX } from '../config/constants'

const { readJsonSync } = fsExtra

const updateBaseLocale = async (replaceObj?: any) => {
  const { baseLangJson } = getJsonPath()
  let baseLangJsonObj

  if (replaceObj) {
    baseLangJsonObj = {}
    const addKeySet = new Set(getKeys(replaceObj))

    for (const k of addKeySet) {
      const v = getValueByKey(replaceObj, k)
      setValueByKey(baseLangJsonObj, k, v)
    }
  }
  else {
    baseLangJsonObj = readJsonSync(baseLangJson.path)
  }
  await writeToJsonFile(path.parse(baseLangJson.path).dir, baseLangJson.name, baseLangJsonObj)
}

const updateLocales = async () => {
  const autoConfig = getAutoConfig()
  const { baseLangJson, otherLangJsons } = getJsonPath()
  const baseLangJsonObj = readJsonSync(baseLangJson.path)

  await updateBaseLocale()

  for (const langJson of otherLangJsons) {
    const langJsonObj = readJsonSync(langJson.path)
    const baseJsonKeySet = new Set(getKeys(baseLangJsonObj))
    const langJsonKeySet = new Set(getKeys(langJsonObj))
    const newObj = sortObjectKey(langJsonObj)
    for (const x of baseJsonKeySet) {
      const baseValue = getValueByKey(baseLangJsonObj, x)
      const langValue = getValueByKey(langJsonObj, x)

      if (langValue === undefined
        || langValue === null
        || langValue === ''
        || isUnTransed(langValue, langJson.name)) {
        let transedWord = ''
        if (autoConfig.transLacaleWord) {
          transedWord = await autoConfig.transLacaleWord(
            baseValue,
            baseLangJson.name,
            langJson.name,
          )
        }
        else {
          transedWord = unTransLocale(baseValue, langJson.name)
        }
        setValueByKey(newObj, x, transedWord)
      }
      langJsonKeySet.delete(x)
    }
    if (langJsonKeySet.size > 0) {
      for (const x of langJsonKeySet) {
        log.info(
          `${x}: ${getValueByKey(langJsonObj, x)} in ${langJson.name
          }.json is deleted, Pleace check it is used in project!`,
        )
      }
    }

    await writeToJsonFile(path.parse(langJson.path).dir, langJson.name, newObj)
  }

  log.success('updateLocales success')
}

const updateLocalesFromXlsx = async (filePath: string) => {
  const res = readXlsxFile(path.join(path.resolve(), filePath))

  const { baseLangJson, otherLangJsons } = getJsonPath()
  if (res) {
    const baseLangJsonObj = readJsonSync(baseLangJson.path)

    for (const langJson of otherLangJsons) {
      const langJsonObj = readJsonSync(langJson.path)
      const typeJsonKeySet = new Set(getKeys(baseLangJsonObj))
      const newJsonObj = sortObjectKey(langJsonObj)

      for (const { data } of res) {
        const keyIndex = data[0].indexOf(KEY_SYMBOL_IN_XLSX)
        const langIndex = data[0].indexOf(langJson.name)

        if (keyIndex !== -1 && langIndex !== -1) {
          for (const aLine of data.slice(1)) {
            if (typeJsonKeySet.has(aLine[keyIndex])) {
              if (aLine[langIndex]) {
                setValueByKey(
                  newJsonObj,
                  aLine[keyIndex],
                  aLine[langIndex].toString(),
                )
              }
            }
          }
        }
      }
      await writeToJsonFile(path.parse(langJson.path).dir, langJson.name, newJsonObj)
      log.success(`Write to Json file success in ${langJson.path}`)
    }
  }
}

export { updateLocalesFromXlsx, updateLocales, updateBaseLocale }
