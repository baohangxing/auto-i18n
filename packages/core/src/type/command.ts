export interface TransCommandOption {
  filePath: string
  generateNewFile: boolean
  newFileName: string
}
