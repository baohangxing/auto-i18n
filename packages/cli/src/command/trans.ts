import path from 'path'
import fs from 'fs'
import type { FileExtension } from '@h1mple/auto-i18n-core'
import {
  KeyCollector, checkInPatterns,
  fgSync, getRecursivePaths, lintFiles, sortObjectKey,
} from '@h1mple/auto-i18n-core'
import type { TransCommandOption } from '../types/config'
import { transform } from '../transform'
import { getAutoConfig, getOutputFileDir } from '../config/config'

import { createFileName } from '../utils/help'
import loger from '../utils/log'
import { updateBaseLocale, updateLocales } from './update'

const trans = async (option: TransCommandOption) => {
  const autoConfig = getAutoConfig()

  const collector = new KeyCollector(autoConfig.baseLocale, loger)

  const transformSingleFile = (filePath: string, replace: boolean) => {
    const autoConfig = getAutoConfig()
    const ext = path.parse(filePath).ext.slice(1) as FileExtension
    const source = fs.readFileSync(filePath, 'utf8')
    const { code } = transform(source, ext, autoConfig.i18nCallRules, collector, loger, replace)
    return code
  }

  let paths: string[] = []

  if (option.transPath)
    paths = getRecursivePaths(option.transPath)
  else
    paths = fgSync(autoConfig.includes)

  collector.init(option.templateFile)
  for (const filePath of paths) {
    const code = transformSingleFile(filePath, option.modifyMode)
    if (option.modifyMode) {
      fs.writeFileSync(filePath, code, 'utf8')
      if (autoConfig.autoFormat && checkInPatterns(filePath, autoConfig.autoFormatRules))
        await lintFiles(filePath)
    }
  }
  if (option.modifyMode) {
    await updateBaseLocale(collector.keyZhMap)
    await updateLocales()
  }
  else {
    const jsonPath = getOutputFileDir(`${createFileName(autoConfig.outputXlsxNameBy.trans)}.json`)

    const obj: any = {}
    for (const x of collector.unExistZhSet)
      obj[x] = x

    fs.writeFileSync(jsonPath, JSON.stringify(sortObjectKey(obj), undefined, 2), 'utf8')
    if (autoConfig.autoFormat && checkInPatterns(jsonPath, autoConfig.autoFormatRules))
      await lintFiles(jsonPath)
  }
}

export { trans }
