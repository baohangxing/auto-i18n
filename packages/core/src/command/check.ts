import path from 'path'
import fs from 'fs'
import fsExtra from 'fs-extra'
import { checkTranslated } from '../check'
import { getAutoConfig, getJsonPath, getOutputFileDir, isUnTransed } from '../config/config'
import type { FileExtension } from '../types'
import { fgSync } from '../utils/glob'
import { writeXlsxFile } from '../utils/excel'
import { createFileName, getKeys, getValueByKey } from '../utils/help'
import log from '../utils/log'

const checkAllTranslated = () => {
  const autoConfig = getAutoConfig()

  const { baseLangJson, otherLangJsons } = getJsonPath()

  let paths: string[] = []

  paths = fgSync(autoConfig.includes)

  const firstRows: string[][] = [
    ['paths', 'untrans key'],
    ['used path', 'keys num', 'keys'],
    ['used key', 'used paths'],
  ]
  const sheetsDatas: string[][][] = [[], [], []]

  const baseLangJsonObj = fsExtra.readJsonSync(baseLangJson.path)

  const allKeysSet = new Set(getKeys(baseLangJsonObj))

  const keysUsedSet = new Set<string>()

  const keysUsedMap = new Map<string, string[]>()

  const langJsonObjMap = {} as any

  firstRows[0].push(baseLangJson.name)

  for (const lang of otherLangJsons) {
    firstRows[0].push(lang.name)
    langJsonObjMap[lang.name] = fsExtra.readJsonSync(lang.path)
  }

  for (const filePath of paths) {
    const autoConfig = getAutoConfig()
    const ext = path.parse(filePath).ext.slice(1) as FileExtension
    const source = fs.readFileSync(filePath, 'utf8')

    const keys: string[] = []
    try {
      keys.push(...checkTranslated(source, ext, autoConfig.i18nCallRules))
    }
    catch (e) {
      log.error(`checkTranslated ${filePath} Error`, e)
      continue
    }

    for (const x of keys) {
      keysUsedSet.add(x)
      keysUsedMap.set(x, [...(keysUsedMap.get(x) ?? []), filePath])
    }
    sheetsDatas[1].push([filePath, keys.length.toString(), keys.join('\n\r')])
  }

  for (const key of keysUsedSet) {
    if (!allKeysSet.has(key))
      log.error(`json do not have key ${key}, pleace check ${baseLangJson.path}`)
    let unTransed = false
    for (const lang of otherLangJsons) {
      const langVal = getValueByKey(langJsonObjMap[lang.name], key)
      if (isUnTransed(langVal, lang.name))
        unTransed = true
    }

    if (unTransed) {
      const item: string[] = [keysUsedMap.get(key)?.join('\n\r') ?? '', key]

      item.push(getValueByKey(baseLangJsonObj, key))

      for (const lang of otherLangJsons) {
        const langVal = getValueByKey(langJsonObjMap[lang.name], key)
        item.push(langVal)
      }

      sheetsDatas[0].push(item)
    }

    sheetsDatas[2].push([key, keysUsedMap.get(key)?.join('\n\r') ?? ''])
  }

  writeXlsxFile(firstRows, ['unTrans key', 'all key detail', 'all key used'], sheetsDatas,
    getOutputFileDir(`${createFileName(autoConfig.outputXlsxNameBy.check)}.xlsx`))

  log.success('check successed')
}

export { checkAllTranslated }
