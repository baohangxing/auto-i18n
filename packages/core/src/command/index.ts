import { Command, Option } from 'commander'
import { CLI_CONFIG_NAME } from '../config/constants'
import type { RevertCommandOption, TransCommandOption } from '../types/config'
import { checkAllTranslated } from '../check'
import { revert } from '../revert'
import { version } from './../../package.json'
import { generateXlsx } from './genXlsx'
import { init } from './init'
import { trans } from './trans'
import { updateLocales, updateLocalesFromXlsx } from './update'

const program = new Command()

program
  .name(CLI_CONFIG_NAME)
  .description('A CLI to help you transform your project '
    + 'to an internationalization project automatically.')
  .version(version)

program.addOption(new Option('-d, --debug').hideHelp())
program.addOption(new Option('-vb, --verbose').hideHelp())

program.on('option:verbose', () => {
  process.env.AUTO_I18N_VERBOSE = program.opts().verbose
})
program.on('option:debug', () => {
  process.env.AUTO_I18N_DEBUG = program.opts().debug
})

program.command('init')
  .description('Init yo-auto-i18n, generate an new default config file automatically.')
  .action(() => {
    init()
  })

program.command('trans')
  .description('Transform a single file or all files.')
  .argument('[transPath]', 'file or directory path')
  .option('--modify', 'whether to modify on the source file')
  .option('-t, --template <string>', 'template json file which inclues all language key')
  .action((transPath, options) => {
    const transCommandOption: TransCommandOption = {
      transPath: transPath ?? '',
      modifyMode: !!options.modify,
      templateFile: options.template ?? '',
    }
    trans(transCommandOption)
  })

program.command('update')
  .description('Update other locales Json files by base locales Json file')
  .option('-t, --templateXlsx <string>', 'update locales json files by the Xlsx template file,'
    + ' the frirt line of the first sheet should have all locales names and the `key`', '')
  .action(async (options) => {
    await updateLocales()
    options.templateXlsx && (await updateLocalesFromXlsx(options.templateXlsx))
  })

program.command('genlsx')
  .description('Generate an Xlsx file by locales json files,'
    + ' the frirt line of in the first sheet of the xlsx will have all locales-name and the `key`')
  .action(() => {
    generateXlsx()
  })

program.command('check')
  .description('Check weather all word has been transformed or not')
  .action(() => {
    checkAllTranslated()
  })

program.command('revert')
  .description('revert a transformed file or files in a directory')
  .argument('<revertPath>', 'file or directory path')
  .option('-t, --target <string>', 'the target language when revert')
  .action(async (revertPath, options) => {
    const revertCommandOption: RevertCommandOption = {
      revertPath: revertPath ?? '',
      target: options.target ?? '',
    }
    await revert(revertCommandOption)
  })

program.parse()
