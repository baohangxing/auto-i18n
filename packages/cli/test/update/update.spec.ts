import path from 'path'
import fs from 'fs'
import { describe, expect, it } from 'vitest'
import { updateLocales } from '../../src/command/update'
import { playground } from '../_helps/playground'

describe('#update', async () => {
  it('should update JSON file', async () => {
    await playground(async ({
      playgroundRoot,
    }) => {
      const jsonPath = path.join(playgroundRoot, 'locales', 'en.json')
      await updateLocales()
      const json = fs.readFileSync(jsonPath).toString()
      expect(json).toMatchSnapshot()
    })
  })
})
