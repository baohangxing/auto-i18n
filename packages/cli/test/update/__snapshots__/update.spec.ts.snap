// Vitest Snapshot v1

exports[`#update > should update JSON file 1`] = `
"{
  \\"dazigao\\": \\"typescript\\",
  \\"title\\": \\"[EN]标题\\"
}
"
`;
